txpool/src/txpool.rs

```diff
@@ -0,0 +1,336 @@
+use std::sync::Arc;
+
+#[derive(Debug, Clone)]
+pub struct TxPool {
+    by_hash: HashMap<TxId, ArcTx>,
+    by_gas_price: PriceSort,
+    by_dependency: Dependency,
+    config: Arc<Config>,
+}
+
+impl TxPool {
+    pub fn new(config: Arc<Config>) -> Self {
+        let max_depth = config.max_depth;
+        Self {
+            by_hash: HashMap::new(),
+            by_gas_price: PriceSort::default(),
+            config,
+            by_dependency: Dependency::new(max_depth),
+        }
+    }
+    pub fn txs(&self) -> &HashMap<TxId, ArcTx> {
+        &self.by_hash
+    }
+
+    pub fn dependency(&self) -> &Dependency {
+        &self.by_dependency
+    }
+
+    // this is atomic operation. Return removed(pushed out/replaced) transactions
+    pub async fn insert(&mut self, tx: ArcTx, db: &dyn TxPoolDB) -> anyhow::Result<Vec<ArcTx>> {
+        if tx.metadata().is_none() {
+            return Err(Error::NoMetadata.into());
+        }
+
+        if self.by_hash.contains_key(&tx.id()) {
+            return Err(Error::NotInsertedTxKnown.into());
+        }
+
+        let mut max_limit_hit = false;
+        // check if we are hiting limit of pool
+        if self.by_hash.len() >= self.config.max_tx {
+            max_limit_hit = true;
+            // limit is hit, check if we can push out lowest priced tx
+            let lowest_price = self.by_gas_price.lowest_price();
+            if lowest_price >= tx.gas_price() {
+                return Err(Error::NotInsertedLimitHit.into());
+            }
+        }
+        // check and insert dependency
+        let rem = self.by_dependency.insert(&self.by_hash, db, &tx).await?;
+        self.by_hash.insert(tx.id(), tx.clone());
+        self.by_gas_price.insert(&tx);
+
+        // if some transaction were removed so we dont need to check limit
+        if rem.is_empty() {
+            if max_limit_hit {
+                //remove last tx from sort
+                let rem_tx = self.by_gas_price.last().unwrap(); // safe to unwrap limit is hit
+                self.remove(&rem_tx);
+                return Ok(vec![rem_tx]);
+            }
+            Ok(Vec::new())
+        } else {
+            // remove ret from by_hash and from by_price
+            for rem in rem.iter() {
+                self.by_hash
+                    .remove(&rem.id())
+                    .expect("Expect to hash of tx to be present");
+                self.by_gas_price.remove(rem);
+            }
+
+            Ok(rem)
+        }
+    }
+
+    /// Return all sorted transactions that are includable in next block.
+    pub fn sorted_includable(&self) -> Vec<ArcTx> {
+        self.by_gas_price
+            .sort
+            .iter()
+            .rev()
+            .map(|(_, tx)| tx.clone())
+            .collect()
+    }
+
+    // When block is updated we need to receive all spend outputs and remove them from txpool
+    pub fn block_update(&self /*spend_outputs: [Input], added_outputs: [AddedOutputs]*/) {
+        todo!()
```

Comment:
Any reason not to implement this now?

Reply:
I wanted to leave it after block inclusion and potential reorgs are flushed out, but when i think about it now we could just use list of inserted Transaction and go with that.
This could be good first task for someone to introduce himself with working's on txpool, having only one person per code module is not the best way. wdyt?